tests/beta/interfaces/test_tape_autograd.py

```diff
@@ -0,0 +1,668 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unit tests for the autograd interface"""
+import pytest
+from pennylane import numpy as np
+
+import pennylane as qml
+from pennylane.beta.tapes import QuantumTape
+from pennylane.beta.queuing import expval, var, sample, probs
+from pennylane.beta.interfaces.autograd import AutogradInterface
+
+
+class TestAutogradQuantumTape:
+    """Test the autograd interface applied to a tape"""
+
+    def test_interface_str(self):
+        """Test that the interface string is correctly identified as autograd"""
+        with AutogradInterface.apply(QuantumTape()) as tape:
+            qml.RX(0.5, wires=0)
+            expval(qml.PauliX(0))
+
+        assert tape.interface == "autograd"
+        assert isinstance(tape, AutogradInterface)
+
+    def test_get_parameters(self):
+        """Test that the get parameters function correctly sets and returns the
+        trainable parameters"""
+        a = np.array(0.1, requires_grad=True)
+        b = np.array(0.2, requires_grad=False)
+        c = np.array(0.3, requires_grad=True)
+        d = np.array(0.4, requires_grad=False)
+
+        with AutogradInterface.apply(QuantumTape()) as tape:
+            qml.Rot(a, b, c, wires=0)
+            qml.RX(d, wires=1)
+            qml.CNOT(wires=[0, 1])
+            expval(qml.PauliX(0))
+
+        assert tape.trainable_params == {0, 2}
+        assert np.all(tape.get_parameters() == [a, c])
+
+    def test_execution(self):
+        """Test execution"""
+        a = np.array(0.1, requires_grad=True)
+        b = np.array(0.2, requires_grad=False)
+
+        def cost(a, b, device):
+            with AutogradInterface.apply(QuantumTape()) as tape:
+                qml.RY(a, wires=0)
+                qml.RX(b, wires=0)
+                expval(qml.PauliZ(0))
+            assert tape.trainable_params == {0}
+            return tape.execute(device)
+
+        dev = qml.device("default.qubit", wires=1)
+        res = cost(a, b, device=dev)
+        assert res.shape == (1,)
+
+    def test_scalar_jacobian(self, tol):
+        """Test scalar jacobian calculation"""
+        a = np.array(0.1, requires_grad=True)
+
+        def cost(a, device):
+            with AutogradInterface.apply(QuantumTape()) as tape:
+                qml.RY(a, wires=0)
+                expval(qml.PauliZ(0))
+            assert tape.trainable_params == {0}
+            return tape.execute(device)
+
+        dev = qml.device("default.qubit", wires=2)
+        res = qml.jacobian(cost)(a, device=dev)
+        assert res.shape == (1,)
+
+        # compare to standard tape jacobian
+        with QuantumTape() as tape:
+            qml.RY(a, wires=0)
+            expval(qml.PauliZ(0))
+
+        tape.trainable_params = {0}
+        expected = tape.jacobian(dev)
+        assert expected.shape == (1, 1)

```

Comment:
How come they have different shapes? :thinking:

Reply:
I could never work this out. As far as I can tell, autograd must be performing a squeeze internally?

Reply:
is it autograd, or is it pennylane?
Probably the difference between returning a scalar array and a scalar?

Reply:
Ah, so I've tracked this down. I think the cause is the `qml.jacobian()` function in `__init__.py`:

```python
    def _jacobian_function(*args, **kwargs):
        argnum = []

        for idx, arg in enumerate(args):
            if getattr(arg, "requires_grad", True):
                argnum.append(idx)

        if len(argnum) == 1:
            return _jacobian(func, argnum[0])(*args, **kwargs)

        return _np.stack([_jacobian(func, arg)(*args, **kwargs) for arg in argnum]).T
```

If `argnum == 1` (as is the case in this test), we should expand the dimension by 1, to match the `argnum > 1` case. Should we mark this as a todo to fix separately? Since this affects the existing code base as well.

```python
        if len(argnum) == 1:
            return np.expand_dims(_jacobian(func, argnum[0])(*args, **kwargs), 1)
```

Reply:
yes definitely